design/garbage-collection.md

```diff
@@ -13,22 +19,54 @@ Yorkie provides garbage collection to solve this problem.
 
 ### Goals
 
-Implements garbage collection system that purges unused nodes not being referenced remotely. 
+Implements garbage collection system that purges unused nodes not being referenced remotely.
 
 ### Non-Goals
 
 For text types, garbage collection works slightly differently: refer to [this document](gc-for-text-type.md).
 
 ## Proposal Details
 
+### Why We Need Version Vector rather than only use Lamport
+Previously we used Lamport (`min_synced_seq`) to handle `Garbage Collection`. It was simple and lightweight to use, but it has crucial weakness that Lamport doesn't guarantee causality.
+![prev-gc-issue](media/prev-gc-issue.jpg)
+
+As you can see from the image above, `min_synced_seq` doesn't guarantee every client receives change.  
+
 ### How It Works
 
 Garbage collection checks that deleted nodes are no longer referenced remotely and purges them completely.
 
-Server records the logical timestamp of the last change pulled by the client whenever the client requests PushPull. And Server returns the smallest logical timestamp, `min_synced_seq` of all clients in response PushPull to the client. `min_synced_seq` is used to check whether deleted nodes are no longer to be referenced remotely or not.
+Server records the version vector of the last change pulled by the client whenever the client requests PushPull. And Server returns the vin version vector, `minVersionVector` of all clients in response PushPull to the client. `minVersionVector` is used to check whether deleted nodes are no longer to be referenced remotely or not.
+
+### What is `minVersionVector`
+Min version vector is the vector which consists of minimum value of every version vector stored in version vector table in database.
+
+Conceptually, min version vector is version vector that is uniformly applied to all users. It ensures that all users have at least this version applied. Which means that it is safe to perform GC.
+
 
-An example of garbage collection:
+```
+// GC algorithm
 
+if (removedAt.lamport <= minVersionVector[removedAt.actor]) {
+    runGC()
+}
+```
+
+```
+// how to find min version vector
+
+min(vv1, vv2) =>
+for (key in vv1 || key in vv2) {
+    minVV[key] = min(vv1[key] || 0, vv2[key] || 0)
+}
+
+ex)
+min([c1:2, c2:3, c3:4], [c1:3, c2:1, c3:5, c4:3])
+=> [c1:2, c2:1, c3:4, c4:0]
+
+```
+### An example of garbage collection:
 #### State 1
 
 ![garbage-collection-1](media/garbage-collection-1.png)
```

Comment:
How is the image's situation created? Is it like this below or just an example situation?

<img width="914" alt="image" src="https://github.com/user-attachments/assets/003c39cc-9800-4623-ae7e-e1f1732db65e">

Reply:
A updates and sends changes to the server, and at the same time, B does the same thing. This situation could happen.